lib/mayaUsd/nodes/layerManager.cpp

```diff
@@ -114,20 +114,23 @@ MStatus disconnectCompoundArrayPlug(MPlug arrayPlug)
     return dgmod.doIt();
 }
 
-/// @brief Verify if the given node is either from the given reference, if that is not null,
-///        or from the main Maya scene if the reference is null.
+/// @brief Verify if the given node is from a reference.
 static bool
-isNodeFromDesiredOrigin(const MFnReference* fromReference, const MFnDependencyNode& node)
+isNodeFromDesiredOrigin(const MFnDependencyNode& node, MayaUsdProxyShapeBase* forProxyShape)
 {
-    if (fromReference) {
-        return fromReference->containsNodeExactly(node.object());
-    } else {
-        return !node.isFromReferencedFile();
-    }
+    const bool proxyIsFromReference
+        = (forProxyShape && MFnDependencyNode(forProxyShape->thisMObject()).isFromReferencedFile());
+    return node.isFromReferencedFile() == proxyIsFromReference;
 }
 
-MayaUsd::LayerManager* findNode(const MFnReference* fromReference = nullptr)
+MayaUsd::LayerManager* findNode(MayaUsdProxyShapeBase* forProxyShape)
 {
+    if (forProxyShape) {
+        if (MayaUsd::LayerManager* layerManager = forProxyShape->getLayerManager()) {
+            return layerManager;
+        }
+    }
+
```

Comment:
This is the main change: the proxy shape node is now able to know which layer manager node to use. Except for existing Mays scene files until they're saved again, so we need the backup option of trying to find the node ourselves. We also need that backup plan in case the previous save had any problem setting the new proxy shape attribute.